File: Week1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    // Get input
    long num;
    do
    {
        num = get_long("Number: ");
    } 
    while (num <= 0);

    int checksum = 0;
    long checknum = num / 10;

    // Calculate checksum
    while (checknum > 0)
    {
        int remainder = checknum % 10 * 2;
        if (remainder < 10)
        {
            checksum += remainder;
        }
        else
        {
            checksum += remainder % 10;
            remainder /= 10;
            checksum += remainder;
        }
        checknum /= 100;
    }

    checknum = num;
    while (checknum > 0)
    {
        checksum += checknum % 10;
        checknum /= 100;
    }
    
    // First check
    if (checksum % 10 != 0)
    {
        printf("INVALID\n");
    }
    else
    {
        // Get the card length and starting digits
        int length = 2;
        while (num >= 100)
        {
            num /= 10;
            length++;
        }

        // Check and print the result
        if ((num == 34 || num == 37) && length == 15)
        {
            printf("AMEX\n");
        }
        else if (num >= 51 && num <= 55 && length == 16)
        {
            printf("MASTERCARD\n");
        }
        else if (num >= 40 && num <= 49 && (length == 13 || length == 16))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
}

File: Week6/credit.py
# TODO
import sys


def main():
    num = input("Number: ")
    length = len(str(num))

    if length != 13 and length != 15 and length != 16:
        print("INVALID")
        sys.exit()

    if length % 2 == 0:
        flag = True
    else:
        flag = False

    checksum = 0
    for i in range(length):
        if flag:
            tmp = int(num[i]) * 2

            if tmp >= 10:
                checksum += tmp - 9
            else:
                checksum += tmp
            flag = False
        else:
            checksum += int(num[i])
            flag = True

    check = int(num[0]) * 10 + int(num[1])
    if checksum % 10 == 0:
        if check == 34 or check == 37:
            print("AMEX")
        elif 51 <= check <= 55:
            print("MASTERCARD")
        elif int(num[0]) == 4:
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()
